Ssr webpack

NataliaTepluhina opened the pull request:
## Description of Problem

This is a draft for SSR documentation with a build step

Deployed demo here: https://deploy-preview-937--vue-docs-next-preview.netlify.app/guide/ssr/introduction.html

Close #533 
Close #726

NataliaTepluhina commented on src/guide/ssr/build-config.md:
```diff
@@ -0,0 +1,83 @@
+# Build Configuration
```
**note**: I still don't like how this file looks like, would be happy to have any improvements suggested

NataliaTepluhina commented:
@skirtles-code @danielroe would you mind checking this draft? It's very raw so far, any suggestions/improvements are much appreciated! 🙇🏻‍♀️

danielroe commented on src/guide/ssr/structure.md:
```diff
@@ -0,0 +1,142 @@
+
+## Avoid Stateful Singletons
+
+When writing client-only code, we are used to the fact that our code will be evaluated in a fresh context every time. However, a Node.js server is a long-running process. When our code is required into the process, it will be evaluated once and stays in memory. This means if you create a singleton object, it will be shared between every incoming request.
```
```suggestion
When writing client-only code, we can assume that our code will be evaluated in a fresh context every time. However, a Node.js server is a long-running process. When our code is first imported by the process, it will be evaluated once and then stay in memory. This means that if you create a singleton object, it will be shared between every incoming request, with the risk of cross-request state pollution.
```

danielroe commented on src/guide/ssr/structure.md:
````diff
@@ -0,0 +1,142 @@
+
+```js
+// app.js
+const { createSSRApp } = require('vue')
+
+module.exports = function createApp() {
````
```suggestion
import { createSSRApp } from 'vue'

export default function createApp() {
```

danielroe commented on src/guide/ssr/structure.md:
```diff
@@ -0,0 +1,142 @@
+
+When writing client-only code, we are used to the fact that our code will be evaluated in a fresh context every time. However, a Node.js server is a long-running process. When our code is required into the process, it will be evaluated once and stays in memory. This means if you create a singleton object, it will be shared between every incoming request.
+
+As seen in the basic example, we are **creating a new root Vue instance for each request.** This is similar to how each user will be using a fresh instance of the app in their own browser. If we use a shared instance across multiple requests, it will easily lead to cross-request state pollution.
+
+So, instead of directly creating an app instance, we should expose a factory function that can be repeatedly executed to create fresh app instances for each request:
```
```suggestion
Therefore, we need to **create a new root Vue instance for each request.** In order to do that, we need to write a factory function that can be repeatedly executed to create fresh app instances for each request:
```

danielroe commented on src/guide/ssr/structure.md:
````diff
@@ -0,0 +1,142 @@
+})
+```
+
+The same rule applies to router and store instances as well. Instead of exporting it directly from a module and importing it across your app, you need to create a fresh instance in `createApp` and inject it from the root Vue instance.
````
```suggestion
The same rule applies to other instances as well (such as the router or store). Instead of exporting the router or store directly from a module and importing it across your app, you should create a fresh instance in `createApp` and inject it from the root Vue instance.
```

danielroe commented on src/guide/ssr/structure.md:
```diff
@@ -0,0 +1,142 @@
+
+## Introducing a Build Step
+
+So far, we haven't discussed how to deliver the same Vue app to the client yet. To do that, we need to use webpack to bundle our Vue app. In fact, we probably want to use webpack to bundle the Vue app on the server as well, because:
+
+- Typical Vue apps are built with webpack and `vue-loader`, and many webpack-specific features such as importing files via `file-loader`, importing CSS via `css-loader` would not work directly in Node.js.
+
+- Although the latest version of Node.js fully supports ES2015 features, we still need to transpile client-side code to cater to older browsers. This again involves a build step.
+
+So the basic idea is we will be using webpack to bundle our app for both client and server - the server bundle will be required by the server and used for SSR, while the client bundle is sent to the browser to hydrate the static markup.
```
```suggestion
So far, we haven't discussed how to deliver the same Vue app to the client yet. To do that, we need to use webpack to bundle our Vue app.

* We need to process the server code with webpack. For example, `.vue` files need to be processed with `vue-loader`, and many webpack-specific features such as importing files via `file-loader` or importing CSS via `css-loader` do not work directly in Node.js.

* Similarly, we need a separate client-side build because although the latest version of Node.js fully supports ES2015 features, older browsers will require the code to be transpiled.

So the basic idea is that we will use webpack to bundle our app for both client and server. The server bundle will be required on the server and used to render static HTML, while the client bundle will be sent to the browser to hydrate the static markup.
```

danielroe reviewed (commented):
This is looking great! I've made a few comments for your consideration.

There is also some legacy ES5 syntax in getting started that was untouched by this PR - it may be worth just making sure everything is consistent across the examples.

skirtles-code reviewed (changes requested):
I haven't tried running the code examples myself yet but I've had a read through and made some suggestions.

I feel that Vue CLI could do more to facilitate this. Relying on a big chunk of webpack configuration in the SSR docs doesn't seem very maintainable. But we are where we are for now...

I should really spend some time learning webpack properly. I can muddle my way through but I don't really have the knowledge required to make good suggestions at the moment.

NataliaTepluhina commented on src/guide/ssr/build-config.md:
````diff
@@ -0,0 +1,83 @@
+
+We will assume you already know how to configure webpack for a client-only project. The config for an SSR project will be largely similar. This example shows a config for Vue CLI project but can be adapted to be used on any webpack build.
+
+```js
+const ManifestPlugin = require('webpack-manifest-plugin')
+const nodeExternals = require('webpack-node-externals')
+const webpack = require('webpack')
````
Agreed! Will restructure the page 👍🏻

NataliaTepluhina commented on src/guide/ssr/routing.md:
```diff
@@ -0,0 +1,129 @@
+
+## Code-Splitting
+
+Code-splitting, or lazy-loading part of your app, helps reducing the amount of assets that need to be downloaded by the browser for the initial render, and can greatly improve TTI (time-to-interactive) for apps with large bundles. The key is "loading just what is needed" for the initial screen.
```
Changed to `size` and applied 👍🏻

NataliaTepluhina commented on src/guide/ssr/routing.md:
````diff
@@ -0,0 +1,129 @@
+
+```js
+// server.js
+const createApp = require('./app')
````
@danielroe will it work fine 🤔 I assumed that this an Express server and it requires ES5 `require` as we're running it as a Node script

NataliaTepluhina commented on src/guide/ssr/structure.md:
````diff
@@ -0,0 +1,142 @@
+
+```js
+// server.js
+const createApp = require('./app')
````
Same note as about about Express server

NataliaTepluhina commented on src/guide/ssr/build-config.md:
```diff
@@ -0,0 +1,83 @@
+
+### Generating `clientManifest`
+
+In addition to the server bundle, we can also generate a client build manifest. With the client manifest and the server bundle, the renderer now has information of both the server _and_ client builds, so it can automatically infer and inject [preload / prefetch directives](https://css-tricks.com/prefetching-preloading-prebrowsing/) and css links / script tags into the rendered HTML.
```
I'll better reword this one 👍🏻

NataliaTepluhina commented on src/guide/ssr/routing.md:
```diff
@@ -0,0 +1,129 @@
+const createApp = require('./app')
+
+server.get('*', async (req, res) => {
+  const { app, router } = await createApp()
```
@skirtles-code we have an asynchronous code in the client entry 🤔  wouldn't it need `await` here?

skirtles-code replied:
I'm getting a bit lost with multiple things having the same name, but as far as I can tell this isn't using the client entry. The `createApp` being called here isn't `async` and doesn't return a promise, so awaiting shouldn't be necessary. I've now got all of this code running locally (not sure I'm doing it right though) and it doesn't seem to need the `await`.

NataliaTepluhina commented on src/guide/ssr/build-config.md:
```diff
@@ -0,0 +1,83 @@
+    webpackConfig.module.rule('vue').uses.delete('cache-loader')
+    webpackConfig.module.rule('js').uses.delete('cache-loader')
+    webpackConfig.module.rule('ts').uses.delete('cache-loader')
+    webpackConfig.module.rule('tsx').uses.delete('cache-loader')
```
Comment added 👍🏻

NataliaTepluhina commented on src/guide/ssr/build-config.md:
```diff
@@ -0,0 +1,83 @@
+
+Notice that in the `externals` option we are whitelisting CSS files. This is because CSS imported from dependencies should still be handled by webpack. If you are importing any other types of files that also rely on webpack (e.g. `*.vue`, `*.sass`), you should add them to the whitelist as well.
+
+If you are using `runInNewContext: 'once'` or `runInNewContext: true`, then you also need to whitelist polyfills that modify `global`, e.g. `babel-polyfill`. This is because when using the new context mode, **code inside a server bundle has its own `global` object.** Since you don't really need it on the server when using Node 7.6+, it's actually easier to just import it in the client entry.
```
Node version removed 👍🏻

NataliaTepluhina commented on src/guide/ssr/routing.md:
```diff
@@ -0,0 +1,129 @@
+
+You may have noticed that our server code uses a `*` handler which accepts arbitrary URLs. This allows us to pass the visited URL into our Vue app, and reuse the same routing config for both client and server!
+
+It is recommended to use the official [vue-router](https://github.com/vuejs/vue-router-next) library for this purpose. Let's first create a file where we create the router. Note that similar to `createRootComponent`, we also need a fresh router instance for each request, so the file exports a `createRouter` function:
```
Removed 👍🏻

NataliaTepluhina commented on src/guide/ssr/hydration.md:
````diff
@@ -0,0 +1,30 @@
+In `entry-client.js`, we are simply mounting the app with this line:
+
+```js
+// this assumes App.vue template root element has `id="app"`
````
Removed 👍🏻

NataliaTepluhina commented on src/guide/ssr/routing.md:
```diff
@@ -0,0 +1,129 @@
+import User from './User.vue'
+
+// to this:
+const User = () => import('./User.vue')
```
Changed to `defineAsyncComponent` here as we mention Vue support for async components

NataliaTepluhina commented on src/guide/ssr/structure.md:
```diff
@@ -0,0 +1,142 @@
+
+## Code Structure with webpack
+
+Now that we are using webpack to process the app for both server and client, the majority of our source code can be written in a universal fashion, with access to all the webpack-powered features. At the same time, there are [a number of things](./universal.md) you should keep in mind when writing universal code.
```
Fixed 👍🏻

NataliaTepluhina commented on src/guide/ssr/structure.md:
```diff
@@ -0,0 +1,142 @@
+    }
+  }
+
+  const app = createSSRApp(rootComponent)
```
Changed to match Vite SSR examples 👍🏻

skirtles-code reviewed (changes requested):
I had a go at running the code in the examples, starting from a fresh CLI project. I've included some notes/suggestions below based on the problems I hit.

NataliaTepluhina commented on src/guide/ssr/routing.md:
````diff
@@ -0,0 +1,117 @@
+
+```js
+// server.js
+const createApp = require('./app')
````
@skirtles-code hmmm, for me it works with `require('./app')` just fine 🤔 I'll leave it like it is for now

NataliaTepluhina replied:
Fixed to be

```
const path = require('path')

const appPath = path.join(__dirname, './dist', 'server', manifest['app.js'])
const createApp = require(appPath).default
```

NataliaTepluhina commented on src/guide/ssr/build-config.md:
````diff
@@ -0,0 +1,89 @@
+Below is a sample `vue.config.js` that adds SSR rendering to a Vue CLI project, but it can be adapted for any webpack build.
+
+```js
+const ManifestPlugin = require('webpack-manifest-plugin')
````
Replaced 👍🏻

NataliaTepluhina commented on src/guide/ssr/routing.md:
````diff
@@ -0,0 +1,129 @@
+const User = () => import('./User.vue')
+```
+
+Note that it is still necessary to use `router.isReady` on both server and client before returning / mounting the app, because the router must resolve async route components ahead of time in order to properly invoke in-component hooks. Let's update our client entry:
````
I've rephrased it and added a link to VueRouter `isReady` API 👍🏻

NataliaTepluhina commented on src/guide/ssr/build-config.md:
```diff
@@ -0,0 +1,89 @@
+    webpackConfig.module.rule('ts').uses.delete('cache-loader')
+    webpackConfig.module.rule('tsx').uses.delete('cache-loader')
+
+    if (!process.env.SSR) {
```
@skirtles-code good point! My code is slightly different but I believe we should extend the guide with the

1) server code
2) package.json scripts to run the build correctly.

We don't need `.env` as we can set `SSR=1` in the build script. I'll add this section to the guide

NataliaTepluhina replied:
Added a server code and a few more changes to the build configuration. including `package.json` scripts 👍🏻

NataliaTepluhina commented on src/guide/ssr/universal.md:
```diff
@@ -0,0 +1,33 @@
+# Writing Universal Code
```
Added :+1

NataliaTepluhina commented:
@skirtles-code @danielroe thank you both for the great review! I added some changes, created a server configuration part and addressed your comments. Would you mind giving a second look to this one?

NataliaTepluhina commented:
@CyberAP I would be really happy if you could take a look too 🙏🏻 I believe your input on SSR is super valuable

CyberAP commented on src/guide/ssr/build-config.md:
```diff
@@ -0,0 +1,107 @@
+
+    webpackConfig
+      .plugin('manifest')
+      .use(new ManifestPlugin({ fileName: 'ssr-manifest.json' }))
```
Must've been a typo.

```suggestion
      .use(new WebpackManifestPlugin({ fileName: 'ssr-manifest.json' }))
```

CyberAP commented on src/guide/ssr/build-config.md:
```diff
@@ -0,0 +1,107 @@
+const webpack = require('webpack')
+
+module.exports = {
+  chainWebpack: webpackConfig => {
```
I was thinking whether it is a good idea or not to have 3 configs here?
One for common rules likes aliases, one for client only and one for server only. Both of them deriving from the base config (using `webpackMerge`).

The reasoning behind this is that using `chainWebpack` for re-configuration is hardly maintainable. It does get very complex extremely fast so noone want to touch this config because noone understands what happens here. I have the same setup on my project and it overwrites a ton of rules exactly like shown here: replacements\deletions inside `if` statements which depend on environment. I think a static webpack config is way more readable because you can understand what it does at the first glance just by the structure of that config (given you have an experience with webpack). This programmatic config is hard to read because you have to parse it line by line working as an interpreter. It is also unclear what the base vue-cli config looks like, so all the `delete` calls look like magic to me.

NataliaTepluhina replied:
I agree with the point but I have a few questions regarding it:
- should we move away from Vue CLI in this case? I am not sure how it's done correctly in the CLI environment as I usually always went with `configureWebpack` / `chainWebpack` in the `vue.config.js

NataliaTepluhina replied:
Added a follow-up to refactor to CLI agnostic build with a base, server and client configs: https://github.com/vuejs/docs-next/issues/969

CyberAP replied:
Yeah, this is probably not feasible with current vue-cli API, so it should be a part of vue-cli agnostic build in that case.

CyberAP commented on src/guide/ssr/build-config.md:
```diff
@@ -0,0 +1,107 @@
+
+    webpackConfig.optimization.splitChunks(false).minimize(false)
+
+    webpackConfig.plugins.delete('hmr')
```
`hmr` is actually critical for a good DX when working in an SSR environment. It is tricky to set it up properly with SSR, so maybe it's not a bad idea to make a separate `vue-webpack-ssr-hmr` package just for that.

NataliaTepluhina replied:
Removed the rule deletion for now and I'll transfer the idea about package to the team 👍🏻

CyberAP commented on src/guide/ssr/universal.md:
```diff
@@ -0,0 +1,33 @@
+
+For browser-only APIs, the common approach is to lazily access them inside client-only lifecycle hooks.
+
+Note that if a 3rd party library is not written with universal usage in mind, it could be tricky to integrate it into an server-rendered app. You _might_ be able to get it working by mocking some of the globals, but it would be hacky and may interfere with the environment detection code of other libraries.
```
I would really like to see some common examples of 3rd party library integrations that do not support SSR evironment. It's probably the most common issue when you're switching from CSR to SSR.

The approaches I would suggest are:
1. Using dynamic imports plus mounted flags (set a flag to `true` when mounted, which triggers dynamic import for components hidden with this flag)
2. Using environment constants: `IS_SERVER`, `IS_CLIENT` and how to define them.

NataliaTepluhina replied:
Good point! For the sake of iteration and to prevent growing this PR further, I am transfering this fo the follow-up issue to improve the guide: https://github.com/vuejs/docs-next/issues/968

CyberAP commented on src/guide/ssr/universal.md:
```diff
@@ -0,0 +1,33 @@
+
+Most [custom directives](/guide/custom-directive.html#custom-directives) directly manipulate the DOM, which will cause errors during SSR. There are two ways to work around this:
+
+1. Prefer using components as the abstraction mechanism and work at the Virtual-DOM level (e.g. using [render functions](/guide/render-function.html#the-dom-tree)) instead.
```
I think the idea here is to replace directives with components as much as possible, but that exact wording is a bit unclear to me unfortunately. It is also unclear how render functions could help me in that case (an example would be much welcomed I think).

NataliaTepluhina replied:
Rephrased a bit and removed a second point (see below)

CyberAP commented on src/guide/ssr/universal.md:
```diff
@@ -0,0 +1,33 @@
+
+1. Prefer using components as the abstraction mechanism and work at the Virtual-DOM level (e.g. using [render functions](/guide/render-function.html#the-dom-tree)) instead.
+
+2. If you have a custom directive that cannot be easily replaced by components, you can provide a "server-side version" of it using the `directives` option when creating the server renderer.
```
I think a link to an example here would be great for quicker migration to SSR.

NataliaTepluhina replied:
As I don't have an example right now (and it will probably take some time to figure it out with the new renderer) I removed this part for now. Will return it back as soon as we have an example 👍🏻

CyberAP commented on src/guide/ssr/build-config.md:
````diff
@@ -0,0 +1,107 @@
+
+```json
+"scripts": {
+  "build:client": "vue-cli-service build --dest dist/client",
````
Maybe it's worth exploring into the `vue-cli` agnostic SSR idea? Because right now if you're not using `vue-cli` your first migration step would be to migrate your current build config to `vue-cli` and only then apply all the SSR practies in the guide.

NataliaTepluhina replied:
Follow-up: https://github.com/vuejs/docs-next/issues/969

CyberAP commented on src/guide/ssr/getting-started.md:
````diff
@@ -64,6 +64,8 @@ yarn add express
 ```
 
 ```js
+// server.js
````
Since most of the people who want to migrate from CSR to SSR would be having a some kind of `main.js` entry file maybe it would make sense to show this migration path as a step-by-step guide how we extract our current code from `main.js` and split it into three different files: `app.js`, `entry-client.js` and `entry-server.js`. The current approach is also very good, but I think it's more focused on writing an SSR app from scracth. Maybe there should be a separate guide for migrating from CSR to SSR? What do you think?

NataliaTepluhina replied:
Agreed, it would be a great improvement! Created a follow-up: https://github.com/vuejs/docs-next/issues/970

CyberAP commented on src/guide/ssr/routing.md:
```diff
@@ -0,0 +1,121 @@
+
+  const appContent = await renderToString(app)
+
+  fs.readFile(path.join(__dirname, '/dist/client/index.html'), (err, html) => {
```
I think this code can be improved by moving the index template reading into some abstract function like `getIndexTemplate`, so we don't mix abstractions and also do some optimisation (right now the file is read on each request which might affect performance).

NataliaTepluhina replied:
Creat catch! Replaced with `readFileSync` and moved out of the server callback :+1:

CyberAP commented on src/guide/ssr/routing.md:
```diff
@@ -0,0 +1,121 @@
+
+const isServer = typeof window === 'undefined'
+
+const history = isServer ? createMemoryHistory() : createWebHistory()
```
I think a slightly better approach here would be to pass the `history` instance on an entry-level, so we have better code-splitting here and won't have to check for `isServer` flag.

CyberAP commented on src/guide/ssr/hydration.md:
```diff
@@ -0,0 +1,35 @@
+
+In development mode, Vue will assert the client-side generated virtual DOM tree matches the DOM structure rendered from the server. If there is a mismatch, it will bail hydration, discard existing DOM and render from scratch. There will be a warning in the browser console but your site will still work.
+
+However, in production mode, this assertion is disabled for maximum performance, which means that **it is crucial you ensure your client- and server-side DOM trees match**. Otherwise your HTML will remain static and non-reactive.
```
I am not sure about this one, does it really fail completely on hydration mismatch? I am not able to test this right now, but I remember that hydration algorithm in Vue 3 was significantly improved so the app doesn't crash completely in case of a mismatch.

danielroe replied:
Good point. I'd need to check this too.

NataliaTepluhina replied:
@CyberAP I still can see a warning in the console for this but in prod mode, it's the same as in dev: app just renders from scratch without crashing 👍🏻 I'll unify both paragraphs removing a distinction for modes

CyberAP commented on src/guide/ssr/routing.md:
```diff
@@ -0,0 +1,121 @@
+# Routing and Code-Splitting
```
As an SSR app developer I am also curious how Critical CSS should be implemeneted correctly with this approach.

NataliaTepluhina replied:
I agree! However, this guide is not full and there are a few missing parts here. Here is an issue with 'what's currently missing': https://github.com/vuejs/docs-next/issues/971

CyberAP commented:
I am also curious whether something about vue-meta integration has changed. In Vue 2 it was important for meta config to be defined only on route config or on a page-level component (a component that can be retrieved from the route config.

Vue 3 also has an improved rendering mechanism which allows for a much flexible data fetching. Previously you must declare your data fetching hooks on a page-level component which significantly limited component composition and data flow. In Vue 3 we have async setup and Suspense that remove that limitation. I think that might also be important to cover.

Lastly, there's a renderToString method in vue server-render package. It produces a lower time to first byte and can increase lighthouse metrics. But it also imposes some restrictions on how you should write your code: no synchronous Teleports (can not replace the markup that has already been sent) or async meta configs (or at least we should wait for the meta to resolve before sending any data, since head tag goes first).

NataliaTepluhina commented:
@CyberAP thank you so much for the great comments! I addressed 'smaller' ones and created follow-ups for the bigger refactorings 👍🏻 Please feel free to add more points here or to follow-up issues, your input is much appreciated

skirtles-code reviewed (changes requested):
A few more suggestions from me.

NataliaTepluhina commented on src/guide/ssr/server.md:
````diff
@@ -0,0 +1,101 @@
+
+- we have to define correct paths to all the assets:
+
+  ```js
+  // server.js
+  server.use(
+    '/img',
+    express.static(path.join(__dirname, './dist/client', 'img'))
+  )
+  server.use('/js', express.static(path.join(__dirname, './dist/client', 'js')))
+  server.use(
+    '/css',
+    express.static(path.join(__dirname, './dist/client', 'css'))
+  )
+  server.use(
+    '/favicon.ico',
+    express.static(path.join(__dirname, './dist/client', 'favicon.ico'))
+  )
+  ```
````
@skirtles-code I prefer to leave it as it is as I believe it's more readable.

NataliaTepluhina commented on src/guide/ssr/structure.md:
````diff
@@ -0,0 +1,143 @@
+}
+```
+
+And our server code now becomes:
````
Fixed the server code

NataliaTepluhina commented on src/guide/ssr/build-config.md:
```diff
@@ -0,0 +1,106 @@
+"scripts": {
+  "build:client": "vue-cli-service build --dest dist/client",
+  "build:server": "SSR=1 vue-cli-service build --dest dist/server",
+  "build": "rm -rf ./dist && npm run build:client && npm run build:server",
```
Removed

NataliaTepluhina commented on src/guide/ssr/routing.md:
```diff
@@ -0,0 +1,121 @@
+
+Code-splitting, or lazy-loading part of your app, helps reduce the size of assets that need to be downloaded by the browser for the initial render, and can greatly improve TTI (time-to-interactive) for apps with large bundles. The key is "loading just what is needed" for the initial screen.
+
+Vue provides [async components support](/guide/component-dynamic-async.html#async-components) with `defineAsyncComponent` method, allowing [webpack to code-split at that point](https://webpack.js.org/guides/code-splitting-async/). All you need to do is:
```
Fixed to use router dynamic import

NataliaTepluhina commented:
@skirtles-code I've addressed your comments, could you please check again?

skirtles-code commented on src/guide/ssr/routing.md:
```diff
@@ -0,0 +1,121 @@
+
+const history = isServer ? createMemoryHistory() : createWebHistory()
+
+const routes: [{ path: '/user', component: MyUser }],
```
```suggestion
const routes = [{ path: '/user', component: MyUser }]
```

skirtles-code commented on src/guide/ssr/routing.md:
````diff
@@ -0,0 +1,121 @@
+```js
+// change this...
+import MyUser from './components/MyUser.vue'
+const routes: [{ path: '/user', component: MyUser }],
+
+// to this:
+routes: [{
+  { path: '/user', component: () => import('./components/MyUser.vue') }
+}]
````
```suggestion
const routes = [{ path: '/user', component: MyUser }]

// to this:
const routes = [
  { path: '/user', component: () => import('./components/MyUser.vue') }
]
```

skirtles-code reviewed (approved):
Last couple of small changes from me and then I think we're good to merge.

NataliaTepluhina commented:
Thanks to everyone who was reviewing this monster! 🙌🏻  I am merging this one and then we'll iterate to shape it through the follow up issues

iGiliy commented:
Excuse me, How to use asyncData in SSR?

iGiliy commented:
> Excuse me, How to use asyncData in SSR?

I tried to use suspense, but <Suspense> is an experimental feature and its API will likely change.
